### ISSUE TITLE
[BUG] Memory allocation exceeds expectations

### ISSUE THREAD

**Issue by mmarouen** [microsoft/DeepSpeed#5484]
@tjruwase
Scenario:
- LORA llama 13b **with gradient checkpointing activated**
- Total number of LORA parameters: **26M**
- Using fp32 precision
- Training on Azure: MPI + pytorch
- HW setup: 256Gb VRAM = 32Gb VRAM x 8 GPUs on a single node
- Training data: ~1M tokens
- deepspeed **zero3 without offload**
- Batch size per device: 32

**Observed behavior**
- All GPUs are busy to 90%
- Training lasts 20h
- Batch size 64: leads to OOM
- Deepspeed reports a memory requirement per GPU ~ 17Gb
- Actual usage per GPU ~30Gb

**Expected behavior**
- Much faster training because capacity seems to be >> memory requirements
- Much larger batch size
Am I missing something? Are the expectations wrong?

**deep speed configuration**
```yaml
{
  "bf16": {
    "enabled": false
  },
  "fp16": {
    "enabled": false,
    "loss_scale": 0,
    "loss_scale_window": 1000,
    "initial_scale_power": 16,
    "hysteresis": 2,
    "min_loss_scale": 1
  },
  "zero_optimization": {
    "stage": 3,
    "overlap_comm": true,
    "contiguous_gradients": true,
    "reduce_bucket_size": 200000000,
    "allgather_bucket_size": 200000000,
    "stage3_prefetch_bucket_size": "auto",
    "stage3_param_persistence_threshold": "auto",
    "sub_group_size": 1000000000,
    "stage3_max_live_parameters": 1000000000,
    "stage3_max_reuse_distance": 1000000000,
    "stage3_gather_16bit_weights_on_model_save": false
  },
  "gradient_accumulation_steps": "auto",
  "gradient_clipping": 1,
  "train_batch_size": "auto",
  "train_micro_batch_size_per_gpu": "auto",
  "wall_clock_breakdown": false
}
```
 **Memory usage snapshots**
```
[2024-04-30 01:43:14,449] [INFO] [utils.py:802:see_memory_usage] Stage 3 initialize beginning
[2024-04-30 01:43:14,449] [INFO] [utils.py:803:see_memory_usage] MA 6.32 GB         Max_MA 7.45 GB         CA 10.05 GB         Max_CA 22 GB 
[2024-04-30 01:43:14,450] [INFO] [utils.py:810:see_memory_usage] CPU Virtual Memory:  used = 23.23 GB, percent = 3.5%
[2024-04-30 01:43:14,454] [INFO] [stage3.py:126:__init__] Reduce bucket size 200000000
[2024-04-30 01:43:14,454] [INFO] [stage3.py:127:__init__] Prefetch bucket size 23592960
[2024-04-30 01:43:14,795] [INFO] [utils.py:802:see_memory_usage] DeepSpeedZeRoOffload initialize [begin]
[2024-04-30 01:43:14,796] [INFO] [utils.py:803:see_memory_usage] MA 6.32 GB         Max_MA 6.32 GB         CA 10.05 GB         Max_CA 10 GB 
[2024-04-30 01:43:14,796] [INFO] [utils.py:810:see_memory_usage] CPU Virtual Memory:  used = 23.23 GB, percent = 3.5%
Parameter Offload: Total persistent parameters: 414720 in 81 params
[2024-04-30 01:43:15,204] [INFO] [utils.py:802:see_memory_usage] DeepSpeedZeRoOffload initialize [end]
[2024-04-30 01:43:15,205] [INFO] [utils.py:803:see_memory_usage] MA 6.24 GB         Max_MA 6.32 GB         CA 10.05 GB         Max_CA 10 GB 
[2024-04-30 01:43:15,205] [INFO] [utils.py:810:see_memory_usage] CPU Virtual Memory:  used = 23.23 GB, percent = 3.5%
[2024-04-30 01:43:15,507] [INFO] [utils.py:802:see_memory_usage] Before creating fp16 partitions
[2024-04-30 01:43:15,508] [INFO] [utils.py:803:see_memory_usage] MA 6.24 GB         Max_MA 6.24 GB         CA 10.05 GB         Max_CA 10 GB 
[2024-04-30 01:43:15,508] [INFO] [utils.py:810:see_memory_usage] CPU Virtual Memory:  used = 23.23 GB, percent = 3.5%
[2024-04-30 01:43:16,224] [INFO] [utils.py:802:see_memory_usage] After creating fp16 partitions: 1
[2024-04-30 01:43:16,226] [INFO] [utils.py:803:see_memory_usage] MA 6.24 GB         Max_MA 6.24 GB         CA 9.94 GB         Max_CA 10 GB 
[2024-04-30 01:43:16,226] [INFO] [utils.py:810:see_memory_usage] CPU Virtual Memory:  used = 23.29 GB, percent = 3.5%
[2024-04-30 01:43:16,505] [INFO] [utils.py:802:see_memory_usage] Before creating fp32 partitions
[2024-04-30 01:43:16,505] [INFO] [utils.py:803:see_memory_usage] MA 6.24 GB         Max_MA 6.24 GB         CA 9.94 GB         Max_CA 10 GB 
[2024-04-30 01:43:16,505] [INFO] [utils.py:810:see_memory_usage] CPU Virtual Memory:  used = 23.29 GB, percent = 3.5%
[2024-04-30 01:43:16,814] [INFO] [utils.py:802:see_memory_usage] After creating fp32 partitions
[2024-04-30 01:43:16,815] [INFO] [utils.py:803:see_memory_usage] MA 6.25 GB         Max_MA 6.25 GB         CA 9.94 GB         Max_CA 10 GB 
[2024-04-30 01:43:16,815] [INFO] [utils.py:810:see_memory_usage] CPU Virtual Memory:  used = 23.29 GB, percent = 3.5%
[2024-04-30 01:43:17,151] [INFO] [utils.py:802:see_memory_usage] Before initializing optimizer states
[2024-04-30 01:43:17,151] [INFO] [utils.py:803:see_memory_usage] MA 6.25 GB         Max_MA 6.25 GB         CA 9.94 GB         Max_CA 10 GB 
[2024-04-30 01:43:17,152] [INFO] [utils.py:810:see_memory_usage] CPU Virtual Memory:  used = 23.25 GB, percent = 3.5%
[2024-04-30 01:43:17,502] [INFO] [utils.py:802:see_memory_usage] After initializing optimizer states
[2024-04-30 01:43:17,502] [INFO] [utils.py:803:see_memory_usage] MA 6.27 GB         Max_MA 6.3 GB         CA 9.94 GB         Max_CA 10 GB 
[2024-04-30 01:43:17,503] [INFO] [utils.py:810:see_memory_usage] CPU Virtual Memory:  used = 23.25 GB, percent = 3.5%
[2024-04-30 01:43:17,503] [INFO] [stage3.py:459:_setup_for_real_optimizer] optimizer state initialized
[2024-04-30 01:43:18,065] [INFO] [utils.py:802:see_memory_usage] After initializing ZeRO optimizer
[2024-04-30 01:43:18,066] [INFO] [utils.py:803:see_memory_usage] MA 7.03 GB         Max_MA 7.03 GB         CA 10.7 GB         Max_CA 11 GB 
[2024-04-30 01:43:18,066] [INFO] [utils.py:810:see_memory_usage] CPU Virtual Memory:  used = 24.42 GB, percent = 3.7%
```
**Actual memory usage**
![gpu_usage](https://github.com/microsoft/DeepSpeed/assets/27000858/65aa1fb9-dea8-4f67-b42b-15bab6b49013)
